addons/calendar/models/calendar.py

```diff
@@ -0,0 +1,1092 @@
+    """
+    Convert a "virtual/recurring event id" (type string) into a real event id (type int).
+    E.g. virtual/recurring event id is 4-20091201100000, so it will return 4.
+    @param calendar_id: id of calendar
+    @param with_date: if a value is passed to this param it will return dates based on value of withdate + calendar_id
+    @return: real event id
+    """
+    if calendar_id and isinstance(calendar_id, (basestring)):
+        res = calendar_id.split('-')
+        if len(res) >= 2:
+            real_id = res[0]
+            if with_date:
+                real_date = time.strftime(DEFAULT_SERVER_DATETIME_FORMAT, time.strptime(res[1], "%Y%m%d%H%M%S"))
+                start = datetime.strptime(real_date, DEFAULT_SERVER_DATETIME_FORMAT)
+                end = start + timedelta(hours=with_date)
+                return (int(real_id), real_date, end.strftime(DEFAULT_SERVER_DATETIME_FORMAT))
+            return int(real_id)
+    return calendar_id and int(calendar_id) or calendar_id
+
+
+def get_real_ids(ids):
+    if isinstance(ids, (basestring, int, long)):
+        t1 = calendar_id2real_id(ids)
+        return t1
+
+    if isinstance(ids, (list, tuple)):
+        t2 = [calendar_id2real_id(id) for id in ids]
+        return t2
+
+
+class CalendarEventType(models.Model):
+    _name = 'calendar.event.type'
+    _description = 'Meeting Type'
+
+    name = fields.Char(required=True)
+
+    _sql_constraints = [('name_uniq', 'unique (name)', "Tag name already exists !")]
+
+
+class CalendarEvent(models.Model):
+    """ Model for Calendar Event """
+    _name = 'calendar.event'
+    _description = "Event"
+    _order = "id desc"
+    _inherit = ["mail.thread", "ir.needaction_mixin"]
+
+    def do_run_scheduler(self):
+        self.env['calendar.alarm_manager'].get_next_mail()
+
+    @api.multi
+    def get_recurrent_date_by_event(self):
+        """Get recurrent dates based on Rule string and all event where recurrent_id is child
+        """
+
+        def todate(date):
+            val = parser.parse(''.join((re.compile('\d')).findall(date)))
+            ## Dates are localized to saved timezone if any, else current timezone.
+            if not val.tzinfo:
+                val = pytz.UTC.localize(val)
+            return val.astimezone(timezone)
+
+        timezone = pytz.timezone(self.env.context.get('tz') or 'UTC')
+        startdate = pytz.UTC.localize(fields.Datetime.from_string(self.start))  # Add "+hh:mm" timezone
+        if not startdate:
+            startdate = datetime.now()
+
+        ## Convert the start date to saved timezone (or context tz) as it'll
+        ## define the correct hour/day asked by the user to repeat for recurrence.
+        startdate = startdate.astimezone(timezone)  # transform "+hh:mm" timezone
+        rset1 = rrule.rrulestr(str(self.rrule), dtstart=startdate, forceset=True)
+        all_events = self.search([('recurrent_id', '=', self.id), '|', ('active', '=', False), ('active', '=', True)])
```

Comment:
you can get records with active=True/False by passing active_test=False in context